src/libsystemd/sd-journal/journal-file.c

```diff
@@ -3107,46 +3107,37 @@ static int generic_array_bisect_plus_one(
                 uint64_t *ret_offset) {
 
         int r;
-        bool step_back = false;
 
         assert(f);
         assert(test_object);
 
         if (n <= 0)
                 return 0;
 
-        /* This bisects the array in object 'first', but first checks
-         * an extra  */
+        /* This bisects the array in object 'first', but first checks an extra. */
         r = test_object(f, extra, needle);
         if (r < 0)
                 return r;
 
-        if (r == TEST_FOUND)
-                r = direction == DIRECTION_DOWN ? TEST_RIGHT : TEST_LEFT;
-
-        /* if we are looking with DIRECTION_UP then we need to first
-           see if in the actual array there is a matching entry, and
-           return the last one of that. But if there isn't any we need
-           to return this one. Hence remember this, and return it
-           below. */
-        if (r == TEST_LEFT)
-                step_back = direction == DIRECTION_UP;
-
-        if (r == TEST_RIGHT) {
-                if (direction == DIRECTION_DOWN)
-                        goto found;
-                else
+        /* If we are looking with DIRECTION_UP then we need to first see if in the actual array there is a
+         * matching entry, and return the last one of that. But if there isn't any we need to return this
+         * one. Hence remember this, and return it below. */
+        if (direction == DIRECTION_DOWN) {
```

Comment:
Can you add a comment for each of the scenarios we handle here as well? I'm having a hard time following